### PR TITLE
Allow adding attribute by Group ID

### PR DESCRIPTION
**Pull request by hahmadia**
#### Summary
Allows you to add attributes via a Group ID.
Pass a GroupIDs key along with an array of group ids you want a attribute to be applied too.
```
    {
                    "Name": ":mattermost: Staff",
                    "UserIDs": ["someuserID3", "someuserID4"],
                    "GroupIDs":["id1","id2"]
     }
```

#### Ticket
No ticket

### CODE REVIEW

File: server/plugin.go [hunk 1]
```diff
@@ -56,14 +56,29 @@ func (p *Plugin) handleGetAttributes(w http.ResponseWriter, r *http.Request) {
...
 			continue
 		}
 		for _, id := range ca.UserIDs {
 			if id == userID {
 				attributes = append(attributes, ca.Name)
 			}
 		}
+		if !contains(attributes, ca.Name) {
+			for _, id := range ca.GroupIDs {
+				usersGroups, err := p.API.GetGroupsForUser(userID)
```
Comment (crspeller) [inline server/plugin.go]: Shouldn't this be called outside the loop?

File: server/plugin.go [hunk 2]
```diff
@@ -56,14 +56,29 @@ func (p *Plugin) handleGetAttributes(w http.ResponseWriter, r *http.Request) {
...
+				if err != nil {
+					http.Error(w, "Could not retrieve groups of this user", http.StatusBadRequest)
+					return
+				}
+				for _, userGroup := range usersGroups {
+					if id == userGroup.Id {
+						attributes = append(attributes, ca.Name)
```
Comment (crspeller) [inline server/plugin.go]: If we moved this to a helper we could return here to save some iterations.

File: server/plugin.go [hunk 3]
```diff
@@ -77,3 +77,21 @@ func (p *Plugin) handleGetAttributes(w http.ResponseWriter, r *http.Request) {
 }
 
 // See https://developers.mattermost.com/extend/plugins/server/reference/
```
Comment (jwilander) [inline server/plugin.go]: I don't think this comment applies to this function

File: server/plugin.go [hunk 4]
```diff
@@ -77,3 +77,21 @@ func (p *Plugin) handleGetAttributes(w http.ResponseWriter, r *http.Request) {
...
+func contains(arr []string, str string, userGroups []*model.Group) bool {
```
Comment (jwilander) [inline server/plugin.go]: This function seems to have two different responsibilities that don't need to be grouped. I'd suggest splitting it into different functions that have specific names for what they're achieving

File: server/plugin.go (hunk 3 above)
Comment (hahmadia) [reply server/plugin.go]: Removed this comment

File: server/plugin.go (hunk 4 above)
Comment (hahmadia) [reply server/plugin.go]: Done! Split into two different functions

File: server/plugin.go [hunk 5]
```diff
@@ -76,4 +76,22 @@ func (p *Plugin) handleGetAttributes(w http.ResponseWriter, r *http.Request) {
...
+func sliceContainsUserGroup(arr []string, userGroups []*model.Group) bool {
+	for _, a := range arr {
+		for _, userGroup := range userGroups {
+			if a == userGroup.Id {
+				return true
+			}
+		}
+	}
+	return false
+}
```
Comment (jwilander) [inline server/plugin.go]: A couple unit tests for the above wouldn't hurt

**Comment by DHaussermann** [comment]
@hahmadia I will take a look at this today. 

Can you please clarify something for me? It's not clear to me where this groups and their membership would be defined. 
Also, would this require a change to the read-me?

**Pull request merged**